File: java/food-ordering/app/src/main/java/dev/restate/sdk/examples/utils/GeoUtils.java
package dev.restate.sdk.examples.utils;

import dev.restate.sdk.examples.types.Location;


public class GeoUtils {
    public static final String DEMO_REGION = "San Jose (CA)";

    private static final double LONG_MIN = -0.0675;
    private static final double LONG_MAX = 0.0675;
    private static final double LAT_MIN = -0.0675;
    private static final double LAT_MAX = 0.0675;
    private static final double SPEED = 0.005;

    public static double randomInInterval(double min, double max){
        double range = max-min;
        return Math.random() * range + min;
    }

    public static Location randomLocation() {
        return new Location(randomInInterval(LONG_MIN, LONG_MAX), randomInInterval(LAT_MIN, LAT_MAX));
    }

    public static Location moveToDestination(Location location, Location destination){
        double dx = destination.lon - location.lon;
        double dy = destination.lat - location.lat;
        double distance = Math.sqrt(dx*dx + dy*dy);
        if (distance < SPEED){
            return destination;
        }
        double ratio = SPEED / distance;
        return new Location(location.lon + dx * ratio, location.lat + dy * ratio);
    }

    public static Long calculateEtaMillis(Location currentLocation, Location targetLocation) {
        var longDiff = Math.abs(targetLocation.lon - currentLocation.lon);
        var latDiff = Math.abs(targetLocation.lat - currentLocation.lat);
        var distance = Math.max(longDiff, latDiff);
        return Math.round(1000 * distance / SPEED);
    }
}

File: end-to-end-applications/kotlin/food-ordering/app/restate-app/src/main/kotlin/dev/restate/sdk/examples/utils/GeoUtils.kt
/*
 * Copyright (c) 2024 - Restate Software, Inc., Restate GmbH
 *
 * This file is part of the Restate examples,
 * which is released under the MIT license.
 *
 * You can find a copy of the license in the file LICENSE
 * in the root directory of this repository or package or at
 * https://github.com/restatedev/examples/
 */
package dev.restate.sdk.examples.utils

import dev.restate.sdk.examples.Location
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.sqrt

object GeoUtils {
  const val DEMO_REGION: String = "San Jose (CA)"

  private const val LONG_MIN = -0.0675
  private const val LONG_MAX = 0.0675
  private const val LAT_MIN = -0.0675
  private const val LAT_MAX = 0.0675
  private const val SPEED = 0.005

  fun randomInInterval(min: Double, max: Double): Double {
    val range = max - min
    return Math.random() * range + min
  }

  fun randomLocation(): Location {
    return Location(randomInInterval(LONG_MIN, LONG_MAX), randomInInterval(LAT_MIN, LAT_MAX))
  }

  fun moveToDestination(location: Location, destination: Location): Location {
    val dx = destination.lon - location.lon
    val dy = destination.lat - location.lat
    val distance = sqrt(dx * dx + dy * dy)
    if (distance < SPEED) {
      return destination
    }
    val ratio = SPEED / distance
    return Location(location.lon + dx * ratio, location.lat + dy * ratio)
  }

  fun calculateEtaMillis(currentLocation: Location, targetLocation: Location): Long {
    val longDiff = abs(targetLocation.lon - currentLocation.lon)
    val latDiff = abs(targetLocation.lat - currentLocation.lat)
    val distance = max(longDiff, latDiff)
    return Math.round(1000 * distance / SPEED)
  }
}
